File: Pain/resources/scripts/lua_types/Events.lua
--- @enum EventType
EventType = {
	Collision = 1,
}

---@class CollisionEvent
---@field a number
---@field b number
---@field normal vec2
---@field penetration number

--- @class Events
Event = {}

---@overload fun(event: EventType, callback: fun(collision_event: CollisionEvent))
---@param event EventType
---@param callback fun(e: table)
function Event.subscribe(event, callback) end

---@overload fun(eventId: integer, callback: fun())
---@param eventId integer
---@param callback fun(eventId, e: table)
function Event.subscribeCustom(eventId, callback) end

---@overload fun(eventId: integer, callback: fun())
---@param eventId integer
---@param t table
function Event.enqueueCustom(eventId, t) end

File: Pain/src/Misc/Events.cpp
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */


/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */


// Events.cpp
#include "Misc/Events.h"
#include "ECS/EventDispatcher.h"
namespace pain
{
enum class EventType : size_t {
#define X(eventclass, eventname) eventname,
  EVENT_TYPE_LIST
#undef X
      Count
};

sol::state &createLuaEventMap(sol::state &lua, reg::EventDispatcher &ed)
{
  lua["Event"] = lua.create_table();
  lua["Event"]["subscribe"] = [&](EventType eventType, sol::function fn) {
    switch (eventType) {
#define X(eventclass, eventname)                                               \
  case EventType::eventname:                                                   \
    ed.subscribe<eventclass>(fn);                                              \
    break;
      EVENT_TYPE_LIST
#undef X
    case EventType::Count:
    default:
      PLOG_W(
          "Warning, I wasn't able to subscribe Event type \"{}\", perhaps "
          "you try to subscribe something custom? E.g. Event.subscribeCustom()",
          static_cast<int>(eventType));
      break;
    }
  };
  lua["Event"]["subscribeCustom"] = [&](const size_t customEventId,
                                        sol::function fn) {
    ed.subscribe(customEventId, fn);
  };

  lua["Event"]["enqueueCustom"] = [&](const size_t customEventId,
                                      const sol::table event) {
    ed.enqueue(customEventId, event);
  };
  lua.new_enum<EventType>("EventType",
                          {
#define X(eventclass, eventname) {#eventname, EventType::eventname}, //
                              EVENT_TYPE_LIST                        //
#undef X                                                             //
                          });
  return lua;
}
#undef EVENT_TYPE_LIST
} // namespace pain
